pkg/distributor/distributor_test.go

```diff
@@ -3729,6 +3729,34 @@ func TestDistributor_UserStats(t *testing.T) {
 			shardSize:      1, // Tenant's shard made of: ingester-zone-a-0, ingester-zone-b-0 and ingester-zone-c-0.
 			expectedSeries: 5,
 		},
+		"multi zone, 5 ingesters, every series successfully replicated to 3 ingesters across 5 zones": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 1, happyIngesters: 1},
+				"zone-b": {numIngesters: 1, happyIngesters: 1},
+				"zone-c": {numIngesters: 1, happyIngesters: 1},
+				"zone-d": {numIngesters: 1, happyIngesters: 1},
+				"zone-e": {numIngesters: 1, happyIngesters: 1},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "series_2", "series_3", "series_4"),
+				},
+				"zone-c": {
+					makeWriteRequest(0, 1, 0, false, false, "series_3", "series_4", "series_5"),
+				},
+				"zone-d": {
+					makeWriteRequest(0, 1, 0, false, false, "series_4", "series_5", "series_1"),
+				},
+				"zone-e": {
+					makeWriteRequest(0, 1, 0, false, false, "series_5", "series_1", "series_2"),
+				},
+			},
+			// We pushed 5 series but every series has been successfully replicated only to 3 ingesters (in different zones).
+			expectedSeries: 4,
```

Comment:
i can't understand why 4 in this case? Shouldn't it be 3 instead? `(3 /*zone-a*/ + 3 /*zone-b*/ + 3 + 3 + 3) / 5 /*replication_factor*/ = 3`

Reply:
I've originally written this test and I confirm what Vladimir is saying. Our count is just an approximation when minimize ingester requests in is use (or even when it's used used but we stop processing ingesters data once quorum is reached).

However, I've the feeling this assertion may be flaky for the test run with `minimizeIngesterRequests: false` because, based on timing conditions, it could also count 5 (case when all ingesters response are processed before we realise quorum is reached and we stop processing response from more ingesters).

Reply:
I've approved the PR but I'm no more sure that this change is an improvement over the current logic. It's probably an improvement when minimise ingesters requests is disabled, but I think it still return the wrong number when minimize is enabled.

Reply:
Was thinking about the problem here: what we do the approximation as follows,

```
MAX(valueAcrossZones) * zonesTotal / RF
```

I think, for multi-zonal deployments, this should return a better approximation for both disable and enable of minimise ingesters requests.